src/common/browser-adapters/browser-event-manager.ts

```diff
@@ -0,0 +1,196 @@
+//      * deferring any events without an associated application-level listener until one is registered
+//   * registering application-level listeners to be called inside the handleEvent listener
+export class BrowserEventManager {
+    private deferredEvents: DeferredEventDetails[] = [];
+    private eventsToApplicationListenersMapping: DictionaryStringTo<ApplicationListener> = {};
+    private eventsToBrowserListenersMapping: DictionaryStringTo<BrowserListener> = {};
+
+    constructor(private readonly promiseFactory: PromiseFactory, private readonly logger: Logger) {}
+
+    public addApplicationListener = (eventType: string, callback: ApplicationListener): void => {
+        if (this.eventsToApplicationListenersMapping[eventType]) {
+            throw new Error(`Listener already registered for ${eventType}`);
+        }
+        this.eventsToApplicationListenersMapping[eventType] = callback;
+        this.processDeferredEvents();
+    };
+
+    public addBrowserListener = (event: Events.Event<any>, eventType: string): void => {
+        const eventListener = this.createBrowserListenerForEventType(eventType);
+        event.addListener(eventListener);
+        this.eventsToBrowserListenersMapping[eventType] = eventListener;
+    };
+
+    public removeListeners(event: Events.Event<any>, eventType: string) {
+        this.removeBrowserListener(event, eventType);
+        this.removeApplicationListener(eventType);
+    }
+
+    private createBrowserListenerForEventType: (eventType: string) => BrowserListener =
+        (eventType: string) =>
+        (...eventArgs: any[]) =>
+            this.handleBrowserEvent(eventType, eventArgs);
+
+    private handleBrowserEvent(eventType: string, eventArgs: any[]): Promise<any> {
+        const responsePromise =
+            this.tryProcessEvent(eventType, eventArgs) ?? this.deferEvent({ eventType, eventArgs });
+
+        return responsePromise.catch(error => {
+            this.logger.error(`Error while processing browser ${eventType} event: `, error);
+
+            // We want to avoid rejecting the Promise we give back to the browser event handler
+            // because doing so is liable to cause the browser to tear down our Service Worker
+            // immediately, even if other events are in progress.
+            return undefined;
+        });
+    }
+
+    private tryProcessEvent(eventType: string, eventArgs: any[]): Promise<any> | null {
+        const applicationListener = this.eventsToApplicationListenersMapping[eventType];
+        return applicationListener
+            ? this.forwardEventToApplicationListener(applicationListener, eventType, eventArgs)
+            : null;
+    }
+
+    private processDeferredEvents(): void {
+        const stillDeferred: DeferredEventDetails[] = [];
+        for (const deferredEvent of this.deferredEvents) {
+            if (deferredEvent.isStale) {
```

Comment:
I think it's preferable to avoid processing an already-timed-out event. If a still-deferred event ever becomes stale, it means that it timed out while it was still in a deferred state, and we've already indicated back to the browser that this event should no longer be preventing the worker from being torn down. If that were to happen, I expect debugging the resulting timeout error log on its own will probably be less misleading than simultaneously debugging a timeout error log plus also whatever exciting result happens when the worker gets torn down mid-handler-execution.

Reply:
(added a comment to this effect)